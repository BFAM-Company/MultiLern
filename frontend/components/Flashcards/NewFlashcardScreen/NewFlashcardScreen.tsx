import React, { useEffect, useState } from 'react';
import { Image, KeyboardAvoidingView, Platform, Pressable, ScrollView, StyleSheet, Text, TextInput, TouchableOpacity, View} from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import { StylesVariables } from '../../../utils/GLOBALS';
import { Controller, useFieldArray, useForm } from 'react-hook-form';

function NewFlashcardScreen({pageSwitcher}: any) {
  const [isEditing, setIsEditing] = useState(false)
  const [title, setTitle] = useState('Tytuł twojego zestawu!')

  const {handleSubmit, watch, control, formState: {errors}} = useForm();
  const { fields, remove, append } = useFieldArray({
    control,
    name: "flashcards",
  });

  useEffect(() => {
    append({});
  }, []);

  const onSubmit = (data: any) => {
    console.log(data);
  };
  
  return (
    <KeyboardAvoidingView
        behavior={Platform.OS === "ios" ? "padding" : "height"}
    >
        <ScrollView>
          <SafeAreaView style={styles.mainContainer}>
            <View style={styles.topicContainer}>
              {isEditing ? 
                <Controller
                  name="title"
                  control={control}
                  render={({field: {onChange, onBlur, value}}) => (
                    <TextInput 
                      style={[styles.topicInput]} 
                      placeholder='Wpisz tytuł'
                      autoFocus
                      onBlur={() => {onBlur; setIsEditing(false)}}
                      onChangeText={(text) => {onChange(text); setTitle(text)}}
                      value={value}
                    />
                  )}
                />
                :
                <Text style={[styles.topicText]} onPress={() => setIsEditing(true)}>{title}</Text>
              } 
              <View style={styles.saveIcon}>
                <Pressable
                  onPress={handleSubmit(onSubmit)}>
                  <Image
                    style={{width: 25, height: 25, tintColor: '#764ba3',}}
                    source= {require('./../../../assets/check-mark_icon.png')}
                  />
                </Pressable>
              </View>
            </View>
            <View style={styles.flashcardsContainer}>
              {fields.map((field, index) => (
                <View key={field.id} style={styles.flashcard}>
                  <View style={styles.closeIcon}>
                    <Pressable
                      onPress={() => remove(index)}
                    >
                      <Image
                        style={{width: 25, height: 25, tintColor: '#764ba3',}}
                        source= {require('./../../../assets/close_icon.png')}
                      />
                    </Pressable>
                  </View>
                  <Controller
                    name={`flashcards[${index}].foreignTranslation`}
                    control={control}
                    render={({field: {onChange, onBlur, value}}) => (
                      <TextInput 
                        style={styles.input}  
                        placeholder='Podaj obce tłumaczenie'
                        onBlur={onBlur}
                        onChangeText={onChange}
                      />
                    )}
                    rules={{required: "Obce tłumaczenie jest wymagane",}}
                    />
                    <Controller
                      name={`flashcards[${index}].polishTranslation`}
                      control={control}
                      render={({field: {onChange, onBlur, value}}) => (
                        <TextInput 
                          style={styles.input}  
                          placeholder='Podaj obce tłumaczenie'
                          onBlur={onBlur}
                          onChangeText={onChange}
                        />
                      )}
                      rules={{required: "Polskie tłumaczenie jest wymagane",}}
                    />
                </View>
              ))}
            </View>
            <View style={styles.addIcon}>
              <Pressable
                onPress={() => append({})}>
                <Image
                  style={{width: 25, height: 25, tintColor: 'white',}}
                  source= {require('./../../../assets/plus_icon.png')}
                />
              </Pressable>
            </View>
          </SafeAreaView>
        </ScrollView>
    </KeyboardAvoidingView>
  );
}

const styles = StyleSheet.create({
    mainContainer: {
      width: '100%',
      height: '100%',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center'
    },
    topicContainer: {
      width: '100%',
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center'
    },
    input: {
      backgroundColor: 'white',
      width: '90%',
      height: StylesVariables.InputSize,
      borderColor: '#764ba3',
      borderWidth: 2,
      borderRadius:20,
      display:'flex',
      flexDirection:'row',
      justifyContent:'space-between',
      alignItems:'center',
      paddingLeft: 10
    },
    topicInput: {
      width: '80%',
      height: StylesVariables.InputSize/1.3,
      borderBottomWidth: 2,
      borderBottomColor: '#c085ff',
      backgroundColor: '#cfa4fc33',
      padding: 0.4,
      color:'#c085ff',
      paddingLeft: 10,
      marginLeft: 10,
    },
    topicText: {
      width: '85%',
      color:'#c085ff',
      fontSize: 26,
      fontWeight:'900',
      textAlign: 'center',
      textAlignVertical: 'center'
    },
    flashcardsContainer: {
      width: '100%',
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      gap: 5,
      marginTop: 10
    },
    flashcard: {
      margin:10,
      width:'90%',
      display:'flex',
      alignItems:'center',
      justifyContent:'space-between',
      gap: 20,
      borderRadius:20,
      padding:20,
      backgroundColor:'#fff',
      shadowColor: "#000",
      shadowOffset: {
              width: 0,
              height: 0,
              // height: -100,
      },
    shadowOpacity: 0.44,
    shadowRadius: 10.32,

    elevation: 16,
      position: 'relative'
    },
    closeIcon: {
      position: 'absolute',
      color: '#764ba3',
      top: 0,
      right: 0,
      zIndex: 2,
      padding: 5
    },
    addIcon: {
      padding: 5,
      width: 50,
      height: 50,
      backgroundColor: '#764ba3',
      borderRadius: 50,
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      margin: 10
    },
    saveIcon: {
      display:  'flex',
      justifyContent: 'center',
      alignItems: 'center',
      padding: 5,
      width: 50,
      height: 50,
      color: '#764ba3',
      zIndex: 2,
      marginRight: 7
    }
})

export default NewFlashcardScreen;