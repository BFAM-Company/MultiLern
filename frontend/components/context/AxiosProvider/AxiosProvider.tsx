import React, { ReactNode, createContext, useContext } from "react";
import axios, { AxiosInstance } from "axios";
import { AuthContext } from "../AuthContext/AuthContext";
import createAuthRefreshInterceptor from "axios-auth-refresh";
import AsyncStorage from "@react-native-async-storage/async-storage";
// import * as Keychain from "react-native-keychain";

interface AuthState {
  accessToken: string | null;
  refreshToken: string | null;
}

interface AxiosContextProps {
  authAxios: AxiosInstance;
  publicAxios: AxiosInstance
}

const AxiosContext = createContext<AxiosContextProps>({
  authAxios: axios.create({
    baseURL: "http://localhost:3001/auth",
  }),
  publicAxios: axios.create({
    baseURL: "http://localhost:3001/",
  })
});

const { Provider } = AxiosContext;

const AxiosProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const authContext = useContext(AuthContext);

  const authAxios = axios.create({
    baseURL: "http://localhost:3001/auth",
  });

  const publicAxios = axios.create({
    baseURL: "http://localhost:3001/",
  });

  authAxios.interceptors.request.use(
    (config) => {
      if (!config.headers.Authorization) {
        config.headers.Authorization = `Bearer ${authContext?.authState.accessToken}`;
      }

      return config;
    },
    (error) => {
      return Promise.reject(error);
    }
  );

  const refreshAuthLogic = (failedRequest: any) => {
    const data = {
      refreshToken: authContext?.authState.refreshToken,
    };

    const options = {
      method: "POST",
      data,
      url: "http://localhost:3001/auth/refresh",
    };

    return axios(options)
      .then(async (tokenRefreshResponse: any) => {
        failedRequest.response.config.headers.Authorization =
          "Bearer " + tokenRefreshResponse.data;

        await AsyncStorage.setItem('auth',JSON.stringify({
            accessToken: tokenRefreshResponse.data,
            refreshToken: authContext?.authState.refreshToken,
        }))

        authContext?.setAuthState({
          ...authContext.authState,
          accessToken: tokenRefreshResponse.data,
        });
        
        return Promise.resolve();
      })
      .catch(async (e: any) => {
        authContext?.setAuthState({
          accessToken: null,
          refreshToken: null,
          authenticated: false
        });
        await AsyncStorage.clear()
      });
  };

  createAuthRefreshInterceptor(authAxios, refreshAuthLogic, {});

  return (
    <Provider
      value={{
        authAxios,
        publicAxios,
      }}
    >
      {children}
    </Provider>
  );
};

export { AxiosContext, AxiosProvider };
