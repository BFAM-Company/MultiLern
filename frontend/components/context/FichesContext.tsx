import React, { createContext, useState, ReactNode } from "react";

// export interface FichesState {
//   id: number
// }

export interface FichesContextType {
  // fichesState: FichesState;
  // setFichesState: (fichesState: FichesState) => void;

  fichesState: number;
  setFichesState: (fichesState: number) => void;
}


const FichesContext = createContext<FichesContextType | null>(null);
const { Provider } = FichesContext;

const FichesProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [fichesState, setFichesState] = useState<number>(0);

  return (
    <Provider
      value={{
        fichesState,
        setFichesState
      }}
    >
      {children}
    </Provider>
  );
};

export { FichesContext, FichesProvider };
