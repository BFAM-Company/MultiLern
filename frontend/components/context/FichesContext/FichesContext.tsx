import React, { createContext, useState, ReactNode } from "react";

export interface FichesContextType {
  fichesState: number;
  setFichesState: (fichesState: number) => void;
}


const FichesContext = createContext<FichesContextType | null>(null);
const { Provider } = FichesContext;

const FichesProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [fichesState, setFichesState] = useState<number>(0);

  return (
    <Provider
      value={{
        fichesState,
        setFichesState
      }}
    >
      {children}
    </Provider>
  );
};

export { FichesContext, FichesProvider };
