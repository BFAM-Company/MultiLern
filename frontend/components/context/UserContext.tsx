import React, { createContext, useState, ReactNode } from "react";

export interface UserDataState {
  nickname: string;
  email: string;
  avatar: any;
  isLogged: boolean
}

export interface UserDataContextType {
  userData: UserDataState | undefined;
  setUserData: (userData: UserDataState) => void;
}


const UserDataContext = createContext<UserDataContextType | null>(null);
const { Provider } = UserDataContext;

const UserDataProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [userData, setUserData] = useState<UserDataState>();

  return (
    <Provider
      value={{
        userData,
        setUserData
      }}
    >
      {children}
    </Provider>
  );
};

export { UserDataContext, UserDataProvider };
