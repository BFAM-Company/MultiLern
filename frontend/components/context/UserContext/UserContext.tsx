import React, { createContext, useState, ReactNode } from "react";

export interface UserDataState {
  id: number,
  nickname: string;
  email: string;
  avatar: any;
}

export interface UserDataContextType {
  userData: UserDataState | undefined;
  setUserData: (userData: UserDataState) => void;
}


const UserDataContext = createContext<UserDataContextType | null>(null);
const { Provider } = UserDataContext;

const UserDataProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [userData, setUserData] = useState<UserDataState>();

  return (
    <Provider
      value={{
        userData,
        setUserData
      }}
    >
      {children}
    </Provider>
  );
};

export { UserDataContext, UserDataProvider };
