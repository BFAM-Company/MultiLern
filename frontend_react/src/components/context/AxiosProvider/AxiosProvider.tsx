import React, { ReactNode, createContext, useContext } from "react";
import axios, { AxiosInstance } from "axios";
import { AuthContext } from "../AuthContext/AuthContext";
import createAuthRefreshInterceptor from "axios-auth-refresh";

interface AuthState {
  accessToken: string | null;
  refreshToken: string | null;
}

interface AxiosContextProps {
  authAxios: AxiosInstance;
  publicAxios: AxiosInstance;
}

// Kontekst Axios
const AxiosContext = createContext<AxiosContextProps>({
  authAxios: axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}/auth`,
    baseURL: `${import.meta.env.VITE_API_URL}/auth`,
  }),
  publicAxios: axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}`,
    baseURL: `${import.meta.env.VITE_API_URL}`,
  }),
});

const { Provider } = AxiosContext;

const AxiosProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const authContext = useContext(AuthContext);

  const authAxios = axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}/auth`,
    baseURL: `${import.meta.env.VITE_API_URL}/auth`,
  });

  const publicAxios = axios.create({
    // baseURL: `${process.env.REACT_APP_API_URL}`,
    baseURL: `${import.meta.env.VITE_API_URL}`,
  });

  // Interceptor dodający nagłówki autoryzacji
  authAxios.interceptors.request.use(
    (config) => {
      if (authContext?.authState.accessToken) {
        config.headers.Authorization = `Bearer ${authContext.authState.accessToken}`;
      }
      return config;
    },
    (error) => Promise.reject(error)
  );

  // Funkcja obsługująca odświeżanie tokenu
  const refreshAuthLogic = (failedRequest: any) => {
    const data = {
      refreshToken: authContext?.authState.refreshToken,
    };

    return axios
      // .post(`${process.env.REACT_APP_API_URL}/auth/refresh`, data)
      .post(`${import.meta.env.VITE_API_URL}/auth/refresh`, data)
      .then((tokenRefreshResponse) => {
        const newAccessToken = tokenRefreshResponse.data;

        // Ustawienie nowego tokenu w localStorage
        localStorage.setItem(
          "auth",
          JSON.stringify({
            accessToken: newAccessToken,
            refreshToken: authContext?.authState.refreshToken,
          })
        );

        // Aktualizacja stanu kontekstu
        authContext?.setAuthState((prevState) => ({
          ...prevState,
          accessToken: newAccessToken,
        }));

        // Ponowienie oryginalnego żądania
        failedRequest.response.config.headers.Authorization = `Bearer ${newAccessToken}`;
        return Promise.resolve();
      })
      .catch(() => {
        // Wylogowanie w przypadku niepowodzenia
        authContext?.setAuthState({
          accessToken: null,
          refreshToken: null,
          authenticated: false,
        });
        localStorage.clear();
      });
  };

  // Dodanie interceptora obsługującego odświeżanie tokenu
  createAuthRefreshInterceptor(authAxios, refreshAuthLogic);

  return (
    <Provider value={{ authAxios, publicAxios }}>
      {children}
    </Provider>
  );
};

export { AxiosContext, AxiosProvider };
