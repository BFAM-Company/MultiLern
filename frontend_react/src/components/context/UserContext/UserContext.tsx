import React, { createContext, useState, ReactNode } from "react";

// Typy danych uÅ¼ytkownika
export interface UserDataState {
  id: number;
  nickname: string;
  email: string;
  avatar: string | null;
}

export interface UserDataContextType {
  userData: UserDataState | undefined;
  setUserData: (userData: UserDataState) => void;
}

// Tworzenie kontekstu
const UserDataContext = createContext<UserDataContextType | null>(null);

const UserDataProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [userData, setUserData] = useState<UserDataState | undefined>(undefined);

  return (
    <UserDataContext.Provider
      value={{
        userData,
        setUserData,
      }}
    >
      {children}
    </UserDataContext.Provider>
  );
};

export { UserDataContext, UserDataProvider };
