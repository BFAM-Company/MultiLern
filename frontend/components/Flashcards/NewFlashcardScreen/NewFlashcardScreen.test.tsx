import React from 'react';
import { render, fireEvent, waitFor, cleanup } from '@testing-library/react-native';
import NewFlashcardScreen from './NewFlashcardScreen';
import { UserDataContext } from '../../context/UserContext/UserContext';
import { AxiosContext } from '../../context/AxiosProvider/AxiosProvider';
import axios from 'axios';
import { Alert, Modal } from 'react-native';
import { FichesContext } from '../../context/FichesContext/FichesContext';

jest.mock('@react-native-async-storage/async-storage', () => require('../../../__mocks__/mock-async-storage'));


afterEach(cleanup);

describe('<NewFlashcardScreen />', () => {
  it('renders correctly', () => {
    const mockPageSwitcher = jest.fn();
    const { toJSON } = render(<NewFlashcardScreen pageSwitcher={mockPageSwitcher} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should create flashcards successfully', async () => {
    const mockUserContext = {
      setUserData: jest.fn(),
      userData: {
        id: 1,
        nickname: 'JohnDoe',
        avatar: 'https://example.com/avatar.jpg',
        email: 'john.doe@example.com',
      }
    };

    const mockAuthAxios = axios.create();
    mockAuthAxios.post = jest.fn().mockResolvedValueOnce({ data: {} });

    const { getByText, getByPlaceholderText, getByTestId } = render(
      <UserDataContext.Provider value={mockUserContext}>
        <AxiosContext.Provider value={{ authAxios: mockAuthAxios, publicAxios: mockAuthAxios }}>
          <NewFlashcardScreen pageSwitcher={() => {}} />
        </AxiosContext.Provider>
      </UserDataContext.Provider>
    );

    fireEvent.press(getByText('Tytuł twojego zestawu!'));
    fireEvent.changeText(getByPlaceholderText('Wpisz tytuł'), 'Test Title');
    fireEvent.changeText(getByPlaceholderText('Podaj obce tłumaczenie'), 'ass');
    fireEvent.changeText(getByPlaceholderText('Podaj polskie tłumaczenie'), 'dupa');

    fireEvent.press(getByTestId('submitButton'));

    await waitFor(() => {
      expect(mockAuthAxios.post).toHaveBeenCalledTimes(1);
      expect(mockAuthAxios.post).toHaveBeenCalledWith(
        '/fiches',
        {
          title: 'Test Title',
          translationsList: [{
            translations: {
                create: {
                    foreignTranslation: 'ass',
                    polishTranslation: 'dupa'
                }
            }
          }],
          userId: 1,
        },
        { headers: { 'Content-Type': 'application/json' } }
      );
    });
  });

  it('should append new flashcard field when add icon is pressed and remove flashcard field when remove icon is pressed', async () => {
    const mockUserContext = {
      setUserData: jest.fn(),
      userData: {
        id: 1,
        nickname: 'JohnDoe',
        avatar: 'https://example.com/avatar.jpg',
        email: 'john.doe@example.com',
      }
    };

    const mockAuthAxios = axios.create();
    mockAuthAxios.post = jest.fn().mockResolvedValueOnce({ data: {} });

    const { getByText, getAllByPlaceholderText, getAllByTestId, getByTestId } = render(
      <UserDataContext.Provider value={mockUserContext}>
        <AxiosContext.Provider value={{ authAxios: mockAuthAxios, publicAxios: mockAuthAxios }}>
          <NewFlashcardScreen pageSwitcher={() => {}} />
        </AxiosContext.Provider>
      </UserDataContext.Provider>
    );

    fireEvent.press(getByTestId('appendButton'))

    await waitFor(() => {
        expect(getAllByPlaceholderText('Podaj polskie tłumaczenie')).toHaveLength(2)
        expect(getAllByPlaceholderText('Podaj obce tłumaczenie')).toHaveLength(2)
    })

    fireEvent.press(getAllByTestId('removeButton')[0])

    await waitFor(() => {
        expect(getAllByPlaceholderText('Podaj polskie tłumaczenie')).toHaveLength(1)
        expect(getAllByPlaceholderText('Podaj obce tłumaczenie')).toHaveLength(1)
    })

  })

  it('should patch flashcards successfully', async () => {
    const mockUserContext = {
        setUserData: jest.fn(),
        userData: {
        id: 1,
        nickname: 'JohnDoe',
        avatar: 'https://example.com/avatar.jpg',
        email: 'john.doe@example.com',
        }
    };

    const mockFichesContext = {
        fichesState: 2,
        setFichesState: jest.fn()
    }

    const mockAuthAxios = axios.create();
    const mockPublicAxios = axios.create();
    mockAuthAxios.patch = jest.fn().mockResolvedValueOnce({ data: {} });
    mockPublicAxios.get = jest.fn().mockResolvedValueOnce({ data: {
        title: 'Test Title',
        fiches_translations: [
            {
                translations: {
                    foreignTranslation: 'ass',
                    polishTranslation: 'dupa',
                }
            }
        ],
        userId: 1,
    }});

    const { getByText, getByPlaceholderText, getByTestId } = render(
        <UserDataContext.Provider value={mockUserContext}>
        <AxiosContext.Provider value={{ authAxios: mockAuthAxios, publicAxios: mockPublicAxios }}>
            <FichesContext.Provider value={mockFichesContext}>
            <NewFlashcardScreen pageSwitcher={() => {}} />
            </FichesContext.Provider>
        </AxiosContext.Provider>
        </UserDataContext.Provider>
    );

    fireEvent.press(getByTestId('submitButton'));

    await waitFor(() => {
        expect(mockAuthAxios.patch).toHaveBeenCalledTimes(1); 
        expect(mockAuthAxios.patch).toHaveBeenCalledWith( 
        `/fiches/${mockFichesContext.fichesState}`, 
        {
            title: 'Tytuł twojego zestawu!',
            translationsList: [],
        },
        {
            headers: {
            'Content-Type': 'application/json'
            }
        }
        );
    });
});

it('should throw NetworkError', async () => {
    const mockUserContext = {
      setUserData: jest.fn(),
      userData: {
        id: 1,
        nickname: 'JohnDoe',
        avatar: 'https://example.com/avatar.jpg',
        email: 'john.doe@example.com',
      }
    };

    const mockAuthAxios = axios.create();
    mockAuthAxios.post = jest.fn().mockRejectedValueOnce({ response: { data: { message: 'Network Error' } } });


    const { getByText, getByPlaceholderText, getByTestId } = render(
      <UserDataContext.Provider value={mockUserContext}>
        <AxiosContext.Provider value={{ authAxios: mockAuthAxios, publicAxios: mockAuthAxios }}>
          <NewFlashcardScreen pageSwitcher={() => {}} />
        </AxiosContext.Provider>
      </UserDataContext.Provider>
    );

    fireEvent.changeText(getByPlaceholderText('Podaj obce tłumaczenie'), 'ass');
    fireEvent.changeText(getByPlaceholderText('Podaj polskie tłumaczenie'), 'dupa');

    fireEvent.press(getByTestId('submitButton'));

    await waitFor(() => {
        expect(mockAuthAxios.post).toHaveBeenCalledTimes(1)
    });
  });

  it('should throw an error when required fields are not filled', async () => {
  const mockUserContext = {
    setUserData: jest.fn(),
    userData: {
      id: 1,
      nickname: 'JohnDoe',
      avatar: 'https://example.com/avatar.jpg',
      email: 'john.doe@example.com',
    }
  };

  const mockAuthAxios = axios.create();

  const { getByText, getByTestId } = render(
    <UserDataContext.Provider value={mockUserContext}>
      <AxiosContext.Provider value={{ authAxios: mockAuthAxios, publicAxios: mockAuthAxios }}>
        <NewFlashcardScreen pageSwitcher={() => {}} />
      </AxiosContext.Provider>
    </UserDataContext.Provider>
  );

  fireEvent.press(getByTestId('submitButton'));

  await waitFor(() => {
    expect(getByText('⚠ Nie uzupełniłes wszystkich pól!')).toBeTruthy();
  });

});

it('should display edit mode when editing the title', async () => {
  const { getByText, getByPlaceholderText, queryByPlaceholderText } = render(<NewFlashcardScreen pageSwitcher={() => {}} />);
  fireEvent.press(getByText('Tytuł twojego zestawu!'));
  fireEvent(getByPlaceholderText('Wpisz tytuł'), 'onBlur');
  await waitFor(() => {
    expect(queryByPlaceholderText('Wpisz tytuł')).toBeNull(); 
  });
});


    

});
