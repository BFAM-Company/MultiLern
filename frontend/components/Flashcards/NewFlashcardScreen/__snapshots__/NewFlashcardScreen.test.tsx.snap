// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewFlashcardScreen /> renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "paddingBottom": 0,
    }
  }
>
  <RCTScrollView>
    <View>
      <RNCSafeAreaView
        edges={
          {
            "bottom": "additive",
            "left": "additive",
            "right": "additive",
            "top": "additive",
          }
        }
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingLeft": 10,
              "width": "100%",
            }
          }
        >
          <Text
            onPress={[Function]}
            style={
              [
                {
                  "color": "#c085ff",
                  "fontSize": 26,
                  "fontWeight": "900",
                  "textAlign": "center",
                  "textAlignVertical": "center",
                  "width": "85%",
                },
              ]
            }
          >
            Tytuł twojego zestawu!
          </Text>
          <View
            style={
              {
                "alignItems": "center",
                "color": "#764ba3",
                "display": "flex",
                "height": 50,
                "justifyContent": "center",
                "marginRight": 7,
                "padding": 5,
                "width": 50,
                "zIndex": 2,
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              testID="submitButton"
            >
              <Image
                source={1}
                style={
                  {
                    "height": 25,
                    "tintColor": "#764ba3",
                    "width": 25,
                  }
                }
              />
            </View>
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "column",
              "gap": 5,
              "marginTop": 10,
              "width": "100%",
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderRadius": 20,
                "display": "flex",
                "elevation": 16,
                "gap": 20,
                "justifyContent": "space-between",
                "margin": 10,
                "padding": 20,
                "position": "relative",
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0.44,
                "shadowRadius": 10.32,
                "width": "90%",
              }
            }
          >
            <View
              style={
                {
                  "color": "#764ba3",
                  "padding": 5,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "zIndex": 2,
                }
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                testID="removeButton"
              >
                <Image
                  source={1}
                  style={
                    {
                      "height": 25,
                      "tintColor": "#764ba3",
                      "width": 25,
                    }
                  }
                />
              </View>
            </View>
            <TextInput
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Podaj obce tłumaczenie"
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "white",
                  "borderColor": "#764ba3",
                  "borderRadius": 20,
                  "borderWidth": 2,
                  "display": "flex",
                  "flexDirection": "row",
                  "height": 93.38000000000001,
                  "justifyContent": "space-between",
                  "paddingLeft": 10,
                  "width": "90%",
                }
              }
            />
            <TextInput
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Podaj polskie tłumaczenie"
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "white",
                  "borderColor": "#764ba3",
                  "borderRadius": 20,
                  "borderWidth": 2,
                  "display": "flex",
                  "flexDirection": "row",
                  "height": 93.38000000000001,
                  "justifyContent": "space-between",
                  "paddingLeft": 10,
                  "width": "90%",
                }
              }
            />
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#764ba3",
              "borderRadius": 50,
              "display": "flex",
              "height": 50,
              "justifyContent": "center",
              "margin": 10,
              "padding": 5,
              "width": 50,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="appendButton"
          >
            <Image
              source={1}
              style={
                {
                  "height": 25,
                  "tintColor": "white",
                  "width": 25,
                }
              }
            />
          </View>
        </View>
      </RNCSafeAreaView>
    </View>
  </RCTScrollView>
</View>
`;
